File: 4. Binary Tree/12_tree_levels.java
import java.util.List;
import java.util.ArrayList;
import java.util.Queue;
import java.util.LinkedList;
import java.util.AbstractMap.SimpleEntry;

class Node<T> {
    T val;
    Node<T> left;
    Node<T> right;

    public Node(T val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

class Source {
    public static <T> List<List<T>> treeLevels(Node<T> root) {
        if (root == null) {
            return List.of();
        }

        List<List<T>> levels = new ArrayList<>();
        Queue<SimpleEntry<Node<T>, Integer>> queue = new LinkedList<>();
        queue.add(new SimpleEntry<>(root, 0));
        while (!queue.isEmpty()) {
            SimpleEntry<Node<T>, Integer> pair = queue.remove();
            Node<T> node = pair.getKey();
            int level = pair.getValue();

            if (level == levels.size()) {
                ArrayList<T> newList = new ArrayList<>();
                newList.add(node.val);
                levels.add(newList);
            } else {
                levels.get(level).add(node.val);
            }

            if (node.left != null) {
                queue.add(new SimpleEntry<>(node.left, level + 1));
            }
            if (node.right != null) {
                queue.add(new SimpleEntry<>(node.right, level + 1));
            }
        }

        return levels;
    }
}

File: 4. Binary Tree/12_tree_levels.py
from collections import deque

class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def tree_levels(root):
    if root is None:
        return []

    levels = []
    queue = deque([(root, 0)])
    while queue:
        node, level = queue.popleft()
        
        if level == len(levels):
            levels.append([node.val])
        else:
            levels[level].append(node.val)

        if node.left:
            queue.append((node.left, level + 1))
        if node.right:
            queue.append((node.right, level + 1))

    return levels
